File: Server/src/common/handler/codec/s2c/Response.cpp
#include "Response.h"

LoginResponse::LoginResponse(const LoginStatus status, const unsigned int userId) :
    RegistrationResponse(ResponseCode::LOGIN, status, userId)
{}

LoginResponse::LoginResponse(const LoginStatus status) :
    LoginResponse(status, -1)
{}

SignupResponse::SignupResponse(const SignupStatus status, const unsigned int userId) :
    RegistrationResponse(ResponseCode::SIGNUP, status, userId),
    succeed(true),
    context("")
{}

SignupResponse::SignupResponse(const SignupStatus status, const std::string &context) :
    RegistrationResponse(ResponseCode::SIGNUP, status, -1),
    succeed(false),
    context(context)
{}

LogoutResponse::LogoutResponse(const LogoutStatus status) :
    ProtocolResponse(ResponseCode::LOGOUT, status)
{}

ErrorResponse::ErrorResponse(const ErrorStatus status, const std::string &message) :
    ProtocolResponse(ResponseCode::ERROR, status),
    message(message)
{}

JoinRoomResponse::JoinRoomResponse(const ConsumingResponseStatus status) :
    ProtocolResponse(ResponseCode::JOIN_ROOM, status)
{}

CreateRoomResponse::CreateRoomResponse(const GenericResponseStatus status, const unsigned int roomId) :
    ProtocolResponse(ResponseCode::CREATE_ROOM, status),
    roomId(roomId)
{}

GetRoomsResponse::GetRoomsResponse(const GenericResponseStatus status, const std::vector<Room*> &rooms) :
    ProtocolResponse(ResponseCode::GET_ROOMS, status),
    rooms(rooms)
{}

GetPlayersInRoomResponse::GetPlayersInRoomResponse(const ConsumingResponseStatus status,
                                                   const std::optional<std::vector<LoggedUser*>> &players) :
    ProtocolResponse(ResponseCode::GET_PLAYERS_IN_ROOM, status),
    players(players)
{}

GetHighScoresResponse::GetHighScoresResponse(const GeneralStatsStatus status, const std::vector<std::pair<std::string, int>> &stats) :
    ProtocolResponse(ResponseCode::GET_HIGH_SCORES, status),
    stats(stats)
{}

GetPersonalStatisticsResponse::GetPersonalStatisticsResponse(const GeneralStatsStatus status, const UserStatistics &stats) :
    ProtocolResponse(ResponseCode::GET_PERSONAL_STATISTICS, status),
    stats(stats)
{}

CloseRoomResponse::CloseRoomResponse(const GenericResponseStatus status) :
    ProtocolResponse(ResponseCode::CLOSE_ROOM, status)
{}

StartGameResponse::StartGameResponse(const GenericResponseStatus status) :
    ProtocolResponse(ResponseCode::START_GAME, status)
{}

LeaveRoomResponse::LeaveRoomResponse(const GenericResponseStatus status) :
    ProtocolResponse(ResponseCode::LEAVE_ROOM, status)
{}

GetRoomStateResponse::GetRoomStateResponse(const GenericResponseStatus protocolStatus, const Room& room)
    : ProtocolResponse(ResponseCode::GET_ROOM_STATE, protocolStatus), room(room)
{}

UpdateRoomDataResponse::UpdateRoomDataResponse(const GenericResponseStatus status) :
    ProtocolResponse(ResponseCode::UPDATE_ROOM_DATA, status)
{}

File: Client/Trivia/Source/Codec/S2C/Response/Response.cs
using System.Collections.Generic;
using Trivia.Models;

namespace Trivia.Codec.S2C.Response;

public enum ResponseCode : byte
{
    Error = 0,
    Login,
    Signup,
    Logout,
    JoinRoom,
    CreateRoom,
    GetRooms,
    GetPlayerInRoom,
    GetHighScores,
    GetPersonalStatistics
}

public enum GenericResponseStatus : uint
{
    Success = 1,
    Error = 0
}

public enum ConsumingResponseStatus  : uint
{
    Success = 1,
    ErrorUnknownResource,
    ErrorInternal = 0
}

/// <typeparam name="TStatus">The enum status type</typeparam>
public abstract record ProtocolResponse<TStatus>(
    TStatus Status
);

/// <typeparam name="TStatus">The enum status type</typeparam>
public abstract record RegistrationResponse<TStatus>(
    TStatus Status,
    int? UserId
) : ProtocolResponse<TStatus>(Status);

public enum LoginStatus : uint
{
    Success = 1,
    FailedInvalidCredentials = 2,
    FailedAlreadyLoggedIn = 3,
    FailedInternalError = 0
}

public record LoginResponse(
    LoginStatus Status,
    int? UserId
) : RegistrationResponse<LoginStatus>(Status, UserId);

public enum SignupStatus : uint
{
    Success = 1,
    FailedUsernameTaken = 2,
    FailedInvalidArgument = 3,
    FailedInternalError = 0
}

public record SignupResponse(
    SignupStatus Status,
    int? UserId = null
) : RegistrationResponse<SignupStatus>(Status, UserId);

public enum LogoutStatus : uint
{
    Success = 1,
    FailedNotLoggedIn = 2,
    FailedInternalError = 0
}

public record LogoutResponse(
    LogoutStatus Status
) : ProtocolResponse<LogoutStatus>(Status);

public enum ErrorStatus : uint
{
    Generic = 0,
    ServerUnimplemented,
    IllegalRequest
}

public record ErrorResponse(
    ErrorStatus Status,
    string Message
) : ProtocolResponse<ErrorStatus>(Status);

//TODO: Add rooms metadata
public record JoinRoomResponse(
    ConsumingResponseStatus Status
) : ProtocolResponse<ConsumingResponseStatus>(Status);

public record CreateRoomResponse(
    GenericResponseStatus Status,
    int RoomId
) : ProtocolResponse<GenericResponseStatus>(Status);


//TODO: When Room is implemented, add below to both.
public record GetRoomsResponse(
    GenericResponseStatus Status
) : ProtocolResponse<GenericResponseStatus>(Status);

public record GetPlayersInRoomResponse(
    GenericResponseStatus Status
) : ProtocolResponse<GenericResponseStatus>(Status);


public enum GenericStatsStatus : uint
{
    Success = 1,
    Error = 0,
    NotConnectedError = 2
}

public record GetHighScoresResponse(
    GenericStatsStatus Status,
    IReadOnlyList<KeyValuePair<string, int>> Stats
) : ProtocolResponse<GenericStatsStatus>(Status);

public record GetPersonalStatisticsResponse(
    GenericStatsStatus Status,
    UserStatistics Stats
) : ProtocolResponse<GenericStatsStatus>(Status);
